platform/EPS.c

```diff
@@ -0,0 +1,97 @@
+/* SPDX-License-Identifier: BSD-2-Clause */
+/*
+ * Copyright (c) 2024, Linaro Limited
+ */
+/* Microsoft Reference Implementation for TPM 2.0
+ *
+ * The copyright in this software is being made available under the BSD
+ * License, included below. This software may be subject to other third
+ * party and contributor rights, including patent rights, and no such
+ * rights are granted under this license.
+ *
+ * Copyright (c) 2018-2023 Microsoft Corporation
+ *
+ * All rights reserved.
+ *
+ * BSD License
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are
+ * met:
+ *
+ * Redistributions of source code must retain the above copyright notice,
+ * this list of conditions and the following disclaimer.
+ *
+ * Redistributions in binary form must reproduce the above copyright
+ * notice, this list of conditions and the following disclaimer in the
+ * documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
+ * IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
+ * TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A
+ * PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED
+ * TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+ * PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+ * LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+ * NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+//
+// Platform Endorsement Primary Seed
+//
+
+#include "TpmError.h"
+#include "Admin.h"
+
+#include <string.h>
+#include <tee_internal_api.h>
+#include <tee_internal_api_extensions.h>
+
+#define TEE_EPS_SIZE      (256/2)   // From TPM2B_RSA_TEST_PRIME in Hierarchy.c
+
+void
+_plat__GetEPS(UINT16 Size, uint8_t *EndorsementSeed)
+{
+    TEE_Result Result = TEE_ERROR_ITEM_NOT_FOUND;
+    uint8_t EPS[TEE_EPS_SIZE] = { 0 };
+    size_t EPSLen;
+
+    IMSG("Size=%" PRIu16 "",Size);
+    IMSG("EPS=%d",TEE_EPS_SIZE);
+
+    pAssert(Size <= (TEE_EPS_SIZE));
+
+    Result = TEE_GetPropertyAsBinaryBlock(TEE_PROPSET_CURRENT_TA,
+                                          "com.microsoft.ta.endorsementSeed",
+                                          EPS,
+                                          &EPSLen);
```

Comment:
If this fTPM implementation is going to be a first-party OP-TEE TA, does it make sense for this 'vendor property' to no longer be vendor-specific?

If I understand correctly, this property needs to be declared per-platform right now, so you'll only get a deterministic EPS with [vexpress](https://github.com/OP-TEE/optee_os/blob/master/core/arch/arm/plat-vexpress/vendor_props.c#L60-L66).

Reply:
It must be a vendor-specific property since it's not defined in the GlobalPlatform specification. We should probably keep this as is since it will cause incompatibility if OP-TEE and TA aren't updated at the same time. This TA can very well be built with an older version of OP-TEE and vice versa.